### DISCUSSION TITLE
How to get chapter title in book project

### DISCUSSION THREAD

**Discussion by jandermoreira** [quarto-dev/quarto-cli#8999] [Q&A]
### Description

I'm working on a filter to get several information from the document to generate an automatic preamble for each chapter in a book

However, I'm stuck in getting the chapter name.

## What I'm trying to do

This is a MWE for the filter:

```lua
-- get_header_1.lua


local function get_header_title(header)
    if header.level == 1 then
        local title = pandoc.utils.stringify(header.content)
        quarto.log.output("Title:", title)
    end

    return header
end

return {
    Header = get_header_title
}
```

If I add this filter to a single document and run `quarto preview single_document.qmd`, I get the debug output in terminal, so the filter finds the title (I'm assuming a single level 1 header in each qmd file). The following document was used for this test and _Title I want is this one_ is part of the output in the terminal.

```qmd
---
filters:
    - get_header_1.lua
---

# Title I want is this one
Text

## Other titles
Text
```

This is the result:

```bash
$ quarto render single_document.qmd --to html
pandoc 
  to: html
  output-file: single_document.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  
Title: Title I want is this one
Output created: single_document.html
```

## The problem
However, when I add this filter to a book project, it prints out the title only for `index.qmd` and ignores all other chapter files.

To reproduce this issue:

#### Create a new book project
Just run `quarto create project book` and choose any name.

#### Add the filter
Copy the above filter to a file `get_header.lua` and add it to the project root.

Also, add the following lines to `_quarto.yml`:

```qmd
filters:
    - get_header.lua
```

Then run quarto:

```sh
$ quarto render --to html
[1/4] index.qmd
Title: Preface
[2/4] intro.qmd
[3/4] summary.qmd
[4/4] references.qmd

Output created: _book/index.html
```

All headers are ignored, except for `index.qmd`.

### Next steps?

I can't imagine other way to the get chapter name using `pandoc` or `quarto` functions in the filter.

If nothing else works, I intend to read the source file and extract the title using regular expressions. This seems to be a not very good looking workaround...

**Comment by mcanouil** [comment]
Note that you could use the "native" format to help you develop your filter, see <https://quarto.org/docs/extensions/lua.html#native-format>.

Why are you using level 1 header rather than the `title` metadata?

```lua
function Meta(meta)
  if meta.title then
    quarto.log.output("Title: " .. pandoc.utils.stringify(meta.title))
  end
end
```

<img width="298" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/272cd174-7234-492e-9992-b9b619c1eb7d">

**Comment by gregmacfarlane** [comment]
I was able to extract the number from the pandoc title string using this filter and function: 

```lua
local function extract_number(str)
  quarto.log.debug("[DEBUG] Input value:", str)

  if type(str) ~= "string" then
      quarto.log.debug("[DEBUG] Invalid type:", type(str))
      return 0
  end

  local num = str:match("(%d+)")
  quarto.log.debug("[DEBUG] Matched substring:", num)

  if num then
      local value = tonumber(num)
      quarto.log.debug("[DEBUG] Converted number:", value)
      return value
  else
      quarto.log.debug("[DEBUG] No number found at start of string.")
      return 0
  end
end

function Meta(meta)
  if meta.title then
    quarto.log.output("Title: " .. pandoc.utils.stringify(meta.title))
    title = pandoc.utils.stringify(meta.title)
    chapternumber = extract_number(title)
    quarto.log.debug("Chapter number: " .. chapternumber )
  end
end
```

Lots of debug statements because I don't really know what I'm doing. @mcanouil, as of Quarto 1.8, the native format doesn't output for book types.

**Reply by mcanouil** [reply]
It never worked for projects, only for individual documents / partial rendering.
Quarto CLI emits a warning for non-supported project formats (since at least 2 years).
